pandas/tests/arrays/test_boolean.py

```diff
@@ -0,0 +1,277 @@
+import operator
+
+import numpy as np
+import pytest
+
+import pandas as pd
+from pandas.arrays import BooleanArray
+from pandas.tests.extension.base import BaseOpsUtil
+import pandas.util.testing as tm
+
+
+def make_data():
+    return [True, False] * 4 + [np.nan] + [True, False] * 44 + [np.nan] + [True, False]
+
+
+@pytest.fixture
+def dtype():
+    return pd.BooleanDtype()
+
+
+@pytest.fixture
+def data(dtype):
+    return pd.array(make_data(), dtype=dtype)
+
+
+def test_boolean_array_constructor():
+    values = np.array([True, False, True, False], dtype="bool")
+    mask = np.array([False, False, False, True], dtype="bool")
+
+    result = BooleanArray(values, mask)
+    expected = pd.array([True, False, True, None], dtype="boolean")
+    tm.assert_extension_array_equal(result, expected)
+
+    with pytest.raises(TypeError):
+        BooleanArray(values.tolist(), mask)
+
+    with pytest.raises(TypeError):
+        BooleanArray(values, mask.tolist())
+
+    with pytest.raises(TypeError):
+        BooleanArray(values.astype(int), mask)
+
+    with pytest.raises(TypeError):
+        BooleanArray(values, None)
+
+
+def test_boolean_array_constructor_copy():
+    values = np.array([True, False, True, False], dtype="bool")
+    mask = np.array([False, False, False, True], dtype="bool")
+
+    result = BooleanArray(values, mask)
+    assert result._data is values
+    assert result._mask is mask
+
+    result = BooleanArray(values, mask, copy=True)
+    assert result._data is not values
+    assert result._mask is not mask
+
+
+@pytest.mark.parametrize(
+    "a, b",
+    [
+        ([True, None], [True, np.nan]),
+        ([None], [np.nan]),
+        ([None, np.nan], [np.nan, np.nan]),
+        ([np.nan, np.nan], [np.nan, np.nan]),
+    ],
+)
+def test_to_boolean_array_none_is_nan(a, b):
+    result = pd.array(a, dtype="boolean")
+    expected = pd.array(b, dtype="boolean")
+    tm.assert_extension_array_equal(result, expected)
+
+
+# @pytest.mark.parametrize(
+#     "values",
+#     [
+#         ["foo", "bar"],
+#         ["1", "2"],
+#         "foo",
+#         [1],
+#         [1.0],
+#         pd.date_range("20130101", periods=2),
+#         np.array(["foo"]),
+#         [[1, 2], [3, 4]],
+#         [np.nan, {"a": 1}],
+#     ],
+# )
+# def test_to_boolean_array_error(values):
+#     # error in converting existing arrays to BooleanArray
+#     with pytest.raises(TypeError):
+#         pd.array(values, dtype="boolean")
```

Comment:
Currently those actually work, as the coercion routine is using numpy coercion: `np.array(['a'], dtype=bool)` actually works and use bool-ness of the values. 
So the question here is if we want to allow this, or rather be more strict (eg only allow actual bools + integer 0/1)? (personally prefer to be strict)

Reply:
Mmm, as long as actual bools and integer 0/1 as accepted then either works for me.

Reply:
Would be nice to have a decision here, and get tests for it. I think we can justify either, since `np.array([np.nan], dtype=bool)` will be `array([True])`. So we're already deviating from NumPy in this limited case.

Do you have a plan for how to do the actual validation? Will it be somewhat expensive?

What are we allowing in `__setitem__`? Do we convert the value to `bool`, or do we require booleans?

Reply:
So an option could be to allow: bool + int/float of 0's and 1's (the float is needed in case it are 0/1's with np.nan), and raise for the rest.

> Do you have a plan for how to do the actual validation? Will it be somewhat expensive?

We can check for actual bool/int/float dtypes (and then for int and float check that the original is equal to the result converted back to numeric; to ensure it were only 0's and 1's). For object we can first do an infer_dtype, and then apply the same logic.

> What are we allowing in __setitem__? Do we convert the value to bool, or do we require booleans?

Currently, we use the same `coerce_to_array` function in `__setitem__` to convert the value(s) to be set to boolean values+mask. So if we restrict like the above, then that follows the same pattern for `__setitem__` (but we could also be more strict in `__setitem__` if that is preferred)